imagetest/test_suites/oslogin/ssh_test.go

```diff
@@ -0,0 +1,322 @@
+	if err != nil {
+		return fmt.Errorf("error getting vm name: %v", err)
+	}
+
+	// Get project and zone of instance.
+	project, zone, err := getProjectZone(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Get instance info.
+	instancesGetReq := &computepb.GetInstanceRequest{
+		Instance: vmname,
+		Project:  project,
+		Zone:     zone,
+	}
+
+	instance, err := client.Get(ctx, instancesGetReq)
+	if err != nil {
+		return fmt.Errorf("error getting instance info: %v", err)
+	}
+	metadata := instance.Metadata
+
+	// Find the key in the metadata. If it doesn't exist, create a new metadata item.
+	found := false
+	for _, item := range metadata.Items {
+		if *(item.Key) == key {
+			item.Value = &value
+			found = true
+			break
+		}
+	}
+	if !found {
+		metadata.Items = append(metadata.Items, &computepb.Items{Key: &key, Value: &value})
+	}
+
+	// Update the metadata on the instance.
+	setMetadataReq := &computepb.SetMetadataInstanceRequest{
+		Instance:         vmname,
+		MetadataResource: metadata,
+		Project:          project,
+		Zone:             zone,
+	}
+	_, err = client.SetMetadata(ctx, setMetadataReq)
+	if err != nil {
+		return fmt.Errorf("error setting metadata: %v", err)
+	}
+	return nil
+}
+
+func sessionOSLoginEnabled(client *ssh.Client) error {
+	// We do not close the session as Run() implicitly closes the session after it's done running.
+	// Otherwise we run into an EOF error.
+	session, err := client.NewSession()
+	if err != nil {
+		return fmt.Errorf("failed to create ssh session: %v", err)
+	}
+	data, err := session.Output("cat /etc/nsswitch.conf")
+	if err != nil {
+		return fmt.Errorf("failed to read /etc/nsswitch.conf: %v", err)
+	}
+
+	for _, line := range strings.Split(string(data), "\n") {
+		line = strings.TrimSpace(line)
+		if strings.HasPrefix(line, "#") {
+			continue
+		}
+		if strings.Contains(line, "passwd:") && !strings.Contains(line, "oslogin") {
+			return fmt.Errorf("OS Login not enabled in /etc/nsswitch.conf.")
+		}
+	}
+
+	return nil
+}
+
+// Checks what's in the /var/google-sudoers.d directory.
+func getSudoFile(client *ssh.Client, sudo string) (string, error) {
+	session, err := client.NewSession()
+	if err != nil {
+		return "", fmt.Errorf("failed to create ssh session: %v", err)
+	}
+
+	modes := ssh.TerminalModes{
+		ssh.ECHO:          0,     // disable echoing
+		ssh.TTY_OP_ISPEED: 14400, // input speed = 14.4kbaud
+		ssh.TTY_OP_OSPEED: 14400, // output speed = 14.4kbaud
+	}
+	if err = session.RequestPty("xterm", 80, 40, modes); err != nil {
+		return "", fmt.Errorf("failed to create pseudo terminal: %v", err)
+	}
+
+	var output []byte
+	in, err := session.StdinPipe()
+	if err != nil {
+		return "", fmt.Errorf("failed to get session stdin: %v", err)
+	}
+	out, err := session.StdoutPipe()
+	if err != nil {
+		return "", fmt.Errorf("failed to get session stdout: %v", err)
+	}
+
+	// Monitor the output stream of the ssh session.
+	go func(in io.WriteCloser, out io.Reader, output *[]byte) {
```

Comment:
I'm having a hard time understanding why this is implemented this way... are we running a remote command? Can't we simply copy the file via ssh?

Reply:
So when we create the SSH client, it turns out we do not have the permissions to access the file. The only way I found to access the `google-sudoers.d` file is to use `sudo`, which requires the user's sudo password to be inputted. As for copying the file, I would imagine we'd run into the same permissions issue, which means we wouldn't be able to copy the file without sudo.

From local testing on test VMs, this was the case -- we needed root permissions to access and read the file.